common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoader.java

```diff
@@ -247,17 +257,76 @@ private void saveSecretKeyToStorage(@NonNull final SecretKey unencryptedKey) thr
+                            span.setStatus(StatusCode.OK);
+                        } catch (Exception ex) {
+                            // 2nd fallback to legacy keygen spec
+                            Logger.error(methodTag, "Second attempt without wrap also failed. Falling back to legacy spec.", ex);
+                            keyPair = generateKeyPairWithLegacySpec(mAlias, keypairGenStartTime);
+                            Span.current().setAttribute(AttributeName.key_pair_gen_successful_method.name(), "legacy_key_gen_spec");
+                            span.setStatus(StatusCode.OK);
+                        }
+                    } else {
+                        Logger.info(methodTag, "Some unknown exception occurred. Running legacy keygen spec logic.");
+                        keyPair = generateKeyPairWithLegacySpec(mAlias, keypairGenStartTime);
+                        Span.current().setAttribute(AttributeName.key_pair_gen_successful_method.name(), "legacy_key_gen_spec");
+                        span.setStatus(StatusCode.OK);
+                    }
+                } catch (final Exception e) {
+                    Logger.warn(methodTag, "Unexpected error with new KeyPairGeneratorSpec. Falling back to legacy spec.");
+                    keyPair = generateKeyPairWithLegacySpec(mAlias, keypairGenStartTime);
+                    Span.current().setAttribute(AttributeName.key_pair_gen_successful_method.name(), "legacy_key_gen_spec");
+                    span.setStatus(StatusCode.OK);
+                } finally {
+                    span.end();
+                }
+            }
+            else {
+                // If flight for using new keygen spec is not enabled, use the legacy spec.
+                Logger.info(methodTag, "Using legacy spec for keypair generation directly.");
+                keyPair = generateKeyPairWithLegacySpec(mAlias, keypairGenStartTime);
+            }
         }
-
         final byte[] keyWrapped = AndroidKeyStoreUtil.wrap(unencryptedKey, keyPair, WRAP_ALGORITHM);
         FileUtil.writeDataToFile(keyWrapped, getKeyFile());
     }
 
+    @RequiresApi(api = Build.VERSION_CODES.P)
+    private KeyPair attemptKeyPairGeneration(@NonNull final String alias, boolean useWrapPurpose, long keypairGenStartTime) throws ClientException{
+        KeyPair keyPair = AndroidKeyStoreUtil.generateKeyPair(
+                WRAP_KEY_ALGORITHM, getSpecForKeyStoreKey(alias, useWrapPurpose));
+        recordKeyGenerationTime(keypairGenStartTime);
+        return keyPair;
+    }
+
+    private KeyPair generateKeyPairWithLegacySpec(@NonNull final String alias, long keypairGenStartTime) throws ClientException{
+        KeyPair keyPair = AndroidKeyStoreUtil.generateKeyPair(
+                WRAP_KEY_ALGORITHM, getLegacySpecForKeyStoreKey(mContext, alias));
+        recordKeyGenerationTime(keypairGenStartTime);
+        return keyPair;
+    }
```

Comment:
The way you are capturing the perf time here (and other places), it means this time is going to represent the total time for key generation (as opposed to just one mechanism). I think this is fine but just want us to be aware of it. On instance where we see poor perf here it could be because the first 2 mechanisms failed and it fell back to 3rd one and so that's a representation of a total time on 3 attempts (vs perhaps just 1 attempt on other devices)